### ISSUE TITLE
Better completion/docs for paramater packs

### ISSUE THREAD

**Issue by njames93** [clangd/clangd#638]
Currently when calling constructors through variadic functions we loose all code completion and documentation for the arguments.
For example using `new Type(^)` we get completions showing all the available args we can pass to Types constructor.
It would be nice if this could be extended to say `std::make_unique<Type>(^)`.

It would be nice if we could populate the completions with the parameters that the parameter pack gets forwarded to.

Doing this automatically would be a huge pain, however it would be much simpler to get this working for a known set of functions(with the possibility to extend through config). I'm suggesting functions from the standard would be ideal for this
`std::make_shared`, `std::make_unique`, `std::<container>::emplace`, `std::<container>::emplace_back`, `std::container::emplace_front` to name but a few.

Down the line it could possibly be extended for improved hover documentation.

**Comment by sam-mccall** [comment]
Yeah this makes sense, we already special case make_unique for the purpose of diagnostics.

There's at least a couple of ways to do this:

 - when the user is typing into a deduced variadic template arg, clang's code completion looks at the AST where it's going to be substituted to find the right completion logic... *and* special case these in clangd so we don't drop the function bodies of the templates... *and* probably deal with issues in standard library implementations where params get forwarded around a lot
 - special-case exactly the "constructor parameters" case: determine when it applies and what the type in question is using heuristics, then apply the constructor completion logic

The latter seems less principled but also maybe way easier...
It wouldn't tirvially generalize to cases like `std::call_once` but I think it would cover the most important ones.

**Comment by HighCommander4** [comment]
This is a slight generalization of #447.

**Comment by PeterTh** [comment]
I think it's worth noting that [this particular feature was implemented in Visual Studio 2019 earlier this year](https://devblogs.microsoft.com/cppblog/intellisense-improvements-in-visual-studio-2019/#perfect-forwarding-intellisense-make_unique-make_shared-emplace-emplace_back) for the common cases.

They also talk about supporting non-standard-library uses of perfect forwarding in the future. I wonder if this should be gated behind some sort of attribute or whether it should always happen when something can be identified as perfect forwarding.

**Comment by upsj** [comment]
I am working on this for inlay hints at https://reviews.llvm.org/D124690

**Comment by SamB** [comment]
> I am working on this for inlay hints at https://reviews.llvm.org/D124690

This was merged as <https://github.com/llvm/llvm-project/commit/a638648fef76146634c2199ce7b90c4bc6bfed01>, btw.
